### DISCUSSION TITLE
Seeking Advice on Using PostgreSQL Triggers for Supabase Authentication in Next.js 13

### DISCUSSION THREAD

**Discussion by vladimirzb** [supabase/supabase#17137] [Questions] [answered]
Hello everyone,

I've been working on a Supabase project that involves authentication. My database structure includes a `Profiles` table that is linked to the `auth.users` table in Supabase. The `Profiles` table serves as a central table that other tables reference via foreign keys. For more context on my database architecture, you can check this [Discussion](https://github.com/orgs/supabase/discussions/17115#discussioncomment-6900929).

I encountered an issue with Row-Level Security (RLS) when trying to insert a new row into the `Profiles` table upon user sign-up. The problem was that the user needs to be signed in to insert into the `Profiles` table, but they aren't signed in yet during the sign-up process.

To solve this, I used PostgreSQL triggers. I created a function that inserts a new row into the `Profiles` table whenever a new user is added to the `auth.users` table. Here's the function definition:

```sql
CREATE OR REPLACE FUNCTION create_profile()
RETURNS TRIGGER AS $$
BEGIN
  INSERT INTO "Profiles" (user_id) VALUES (NEW.id);
  RETURN NEW;
END;
$$ LANGUAGE plpgsql;
```

I then set up a trigger to call this function whenever a new row is inserted into the `auth.users` table.

Here's the relevant part of my sign-up code in my Next.js 13 project:

```javascript
  const handleSignUp = async () => {
    const response = await supabase.auth.signUp({
      email,
      password,
      options: {
        emailRedirectTo: `${location.origin}/auth/callback`,
      },
    });
  
    // Log the sign-up response
    console.log("Sign-up response:", JSON.stringify(response, null, 2));
  
    if (response.data && response.data.user) {
      // Log the identities array
      console.log("Identities array:", JSON.stringify(response.data.user.identities, null, 2));
  
      if (response.data.user.identities && response.data.user.identities.length > 0) {
        console.log("Sign-up successful!");
      } else {
        console.log("Email address is already taken.");
        const signInResponse = await supabase.auth.signInWithPassword({
          email,
          password,
        });
  
        if (signInResponse.error) {
          console.error("An error occurred during sign-in:", signInResponse.error.message);
        } else {
          console.log("Successfully signed in existing user!");
        }
      }
    } else {
      console.error("An error occurred during sign-up:", response.error?.message);
    }
  
    router.refresh();
  };
```

Interestingly, when I manually created a user in the `auth.users` table via SQL, the trigger worked as expected. However, when I tried to sign up a user through my Next.js 13 project, not only did I encounter the following error message, but the new account in `auth.users` also wasn't created:

```text
An error occurred during sign-up: Database error saving new user
```

Is this a good solution to the problem I was facing? Are there any potential issues or better approaches that I should consider? Why might the trigger work in one context but not the other?

Thank you for your time and insights!

Best regards,
Vladimir

**Comment by GaryAustin1** [comment, marked as answer]
Also answered you on Discord...
Please look at the user management starter guide/template in the SQL editor of the dashboard.  Your function needs to be security definer type.  Also if you go to the logs on the dashboard and look in Postgres log you should see your specific error.

**Reply by vladimirzb** [reply]
Thank you for your patience and assistance! Being new to PostgreSQL and Supabase, I've found this learning curve quite steep, but your guidance has been invaluable. The user management starter guide you recommended, specifically [this section](https://supabase.com/docs/guides/auth/managing-user-data), was exactly what I needed to resolve my issue. I've successfully implemented the 'security definer' type for my function, and everything is working as expected now. Thanks again for your timely and insightful help!
